### PR TITLE
MAINT: Fix sign-compare warnings in mem_overlap.c.

### PR DESCRIPTION
**Pull request by charris**

### CODE REVIEW

**Comment by charris** [comment]
@pv Heads up.  Perhaps defining some local types would help in tracking the required types.

File: numpy/core/src/private/mem_overlap.c [hunk 1]
```diff
@@ -797,12 +799,12 @@ solve_may_share_memory(PyArrayObject *a, PyArrayObject *b,
        the extent check above.)
     */
 
-    rhs = MIN(end2 - 1 - start1, end1 - 1 - start2);
-
-    if (rhs != (npy_uintp)rhs) {
```
Comment (charris) [inline numpy/core/src/private/mem_overlap.c]: I think this was a genuine bug.

File: numpy/core/src/private/mem_overlap.c [hunk 2]
```diff
@@ -711,7 +711,7 @@ static int
 strides_to_terms(PyArrayObject *arr, diophantine_term_t *terms,
                  unsigned int *nterms, int skip_empty)
```
Comment (eric-wieser) [inline numpy/core/src/private/mem_overlap.c]: Seems `nterms` should be an `int`, to match `PyArray_NDIM`

File: numpy/core/src/private/mem_overlap.c [hunk 3]
```diff
@@ -845,8 +847,7 @@ solve_may_have_internal_overlap(PyArrayObject *a, Py_ssize_t max_work)
 {
     diophantine_term_t terms[NPY_MAXDIMS+1];
     npy_int64 x[NPY_MAXDIMS+1];
-    unsigned int nterms;
-    int i, j;
+    unsigned int i, j, nterms;
```
Comment (eric-wieser) [inline numpy/core/src/private/mem_overlap.c]: Which would make these `int` again

Comment (charris) [reply numpy/core/src/private/mem_overlap.c]: That would probably work, but `nterms` gets passed down through several routines and I'd need to audit all of them to make sure nothing could go wrong. This is the safest, quick fix. I do think the overlap code could use a more exhaustive cleanup now that it has had time to settle, but I don't want to do it in this PR.

Comment (charris) [reply numpy/core/src/private/mem_overlap.c]: My guess is that no unsigned integers are needed in the code, except possibly `npy_uintp`, and I'm not sure about that as `ptrdiff_t` should serve if present. I'll open an issue for a `mem_overlap` review.

File: numpy/core/src/private/mem_overlap.c [hunk 4]
```diff
@@ -415,7 +415,7 @@ diophantine_dfs(unsigned int n,
             x[0] = x1 + c1*t_l;
             x[1] = x2 - c2*t_l;
             if (require_ub_nontrivial) {
-                int j, is_ub_trivial;
+                unsigned int j, is_ub_trivial;
```
Comment (eric-wieser) [inline numpy/core/src/private/mem_overlap.c]: Using `unsigned` for `is_ub_trivial` is a little odd - would prefer to leave it as `int`, or switch to `npy_bool`

File: numpy/core/src/private/mem_overlap.c (hunk 3 above)
Comment (charris) [reply numpy/core/src/private/mem_overlap.c]: See #10296.

File: numpy/core/src/private/mem_overlap.c (hunk 4 above)
Comment (charris) [reply numpy/core/src/private/mem_overlap.c]: Yes, it only takes the values 0 and 1. However, see above about major refactoring of the code.

Comment (eric-wieser) [reply numpy/core/src/private/mem_overlap.c]: This particular one seems like a step in the wrong directions, without the justification of fixing any warnings.

But I suppose I'm being overly picky - as a whole, the patch is better than doing nothing.

Comment (charris) [reply numpy/core/src/private/mem_overlap.c]: OK, I made `is_ub_trivial` an integer.

Comment (charris) [reply numpy/core/src/private/mem_overlap.c]: It's not that I disagree with you, it's that in my past experience with doing number theoretic stuff with fixed sized integers lots of things can go wrong, so I regard this code as likely somewhat delicate. Enough so, that I wouldn't want to make extensive changes without thoroughly understanding the algorithm, which is why I pinged @pv.

**Comment by eric-wieser** [comment]
Oops, was intending to come back and merge this but forgot!

**Pull request merged**